### PR TITLE
Improvements in CreateUserForm

### LINKED ISSUE #9733: Required Fields Error and Unnecessary Network Requests
- `Submit` button should be disabled using `isDirty` property from react-hook-form in Update Encounter form to prevent unnecessary network requests

Image

Image

- Date of Birth can't be in future - update zod schema
- Phone number field should not take more than 10 numbers after `+91`
- Have a red asterisk in front of the label of all required fields just like `State` field.

To Reproduce
Click [here](https://care.ohc.network/facility/1dc493a0-954f-4181-b18f-e3e5bc1086f0/patient/0029dfaa-9c0f-4317-aab8-f2f17d8b0fe1/encounter/05bb8cc1-8329-4eb8-986b-5a5bbb3bc01a/edit_encounter)

**Expected behavior**
Submit button should be disabled if no changes are made to form

### PR DESCRIPTION
**Pull request by rajku-dev**
## Proposed Changes

- Fixes #9733
- Continues #9749 
- Use DatePicker instead
- Create new UserCreateRequest Type
- Modify Schema

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Release Notes

- **New Features**
  - Enhanced form validation with detailed error messages for user registration.
  - Added more gender options in user profile.
  - Improved date picker with advanced date selection controls.
  - New optional property for disabling specific dates in the date picker.

- **Localization**
  - Added comprehensive validation messages for date of birth and phone number.

- **User Experience**
  - More robust input validation for required fields.
  - Real-time form validation feedback.
  - Loading state indicators during form submission.
  - Enhanced error handling and success notifications in the user creation form.

### CODE REVIEW

**Comment by rajku-dev** [comment]
@Jacobjeevan please review the changes , last pr was closed due to so many merge [conflicts](https://github.com/ohcnetwork/care_fe/commit/86eba9b0860e16a6392eb030fa7fcf36557b2e04)

File: src/types/user/user.ts [hunk 1]
```diff
@@ -14,3 +14,20 @@ export type UserBase = {
...
+  gender: (typeof GENDER_TYPES)[number]["id"];
+  password: string;
+  geo_organization: string;
+  username: string;
+  first_name: string;
+  last_name: string;
+  email: string;
+  phone_number: string;
+  alt_phone_number?: string;
+  date_of_birth: Date;
```
Comment (rithviknishad) [inline src/types/user/user.ts]: how does Date object go through the JSON request body?

File: src/components/Users/CreateUserForm.tsx [hunk 2]
```diff
@@ -365,9 +403,15 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
+                <FormLabel required>{t("date_of_birth")}</FormLabel>
                 <FormControl>
-                  <Input type="date" {...field} />
+                  <DatePicker
+                    date={field.value}
+                    onChange={(date) => field.onChange(date)}
+                    disabled={(date) =>
+                      date > new Date() || date < new Date("1900-01-01")
+                    }
+                  />
```
Comment (rithviknishad) [inline src/components/Users/CreateUserForm.tsx]: Why was this changed to `DatePicker`? Try entering a realistic date of birth. It's really hard for the users to click thousand times to reach the desired month

**Review by rithviknishad — CHANGES_REQUESTED** [review]
do a self test; the button is disabled even after filling all the fields, nor are there any errors shown in the form.

image

File: src/components/Users/CreateUserForm.tsx (hunk 2 above)
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: changed it back to input.

File: src/types/user/user.ts (hunk 1 above)
Comment (rajku-dev) [reply src/types/user/user.ts]: fixed to string

**Comment by rajku-dev** [comment]
> do a self test; the button is disabled even after filling all the fields, nor are there any errors shown in the form.
> 
> image

it was due to `.refine` not running on password `onChange` 
moved the form setting logic inside existing `useEffect
 All edge cases have been handled, please review.

**Comment by rajku-dev** [comment]
> do a self test; the button is disabled even after filling all the fields, nor are there any errors shown in the form.

 it was due to `.refine` not running on password `onChange`,
 moved the form setting logic inside existing `useEffect`
 All edge cases have been handled, please review.

https://github.com/user-attachments/assets/14bd3916-88bf-4ab9-b7d5-9969cf390e58

**Comment by rithviknishad** [comment]
let's not move it to useEffect. let's stick with refine itself. however, refine would get triggered when form is attempted to submit right? so what's the issue? why was the button disabled in the first place which prevented submitting it?

**Comment by rajku-dev** [comment]
> let's not move it to useEffect. let's stick with refine itself. however, refine would get triggered when form is attempted to submit right? so what's the issue? why was the button disabled in the first place which prevented submitting it?
image


https://github.com/user-attachments/assets/398bc9ad-72fc-48fb-b951-0db14b7c5f86


.refine is not triggering on changing the password fields, so i had to apply the hack

**Comment by AdityaJ2305** [comment]
> refine would get triggered when form is attempted to submit right ? 

@rajku-dev, I think you missed this line from rithvik's comment

**Comment by rajku-dev** [comment]
> > refine would get triggered when form is attempted to submit right ? 
> 
> @rajku-dev, I think you missed this line from rithvik's comment 

 The problem is, password has mismatch error that makes the submit button disable(since we are doing invalid check), but not showing under input field
I have demonstrated in video please look carefully, that explains everything

**Comment by rithviknishad** [comment]
The real reason why refine doesn't get triggered is, because there are other errors that was raised "before" refine was executed, that needs to be solved first in order for refine to be executed.

Check what those other errors are. 

Refer:

- https://github.com/orgs/react-hook-form/discussions/8516
- https://github.com/orgs/react-hook-form/discussions/8809

**Comment by rajku-dev** [comment]
> The real reason why refine doesn't get triggered is, because there are other errors that was raised "before" refine was executed, that needs to be solved first in order for refine to be executed.
> 
> Check what those other errors are.

i found that .refine will trigger only when full data is passed to it (when trying to submit), 
i also tried to console log, it was not logging any password mismatch errors
image

File: src/components/Users/CreateUserForm.tsx [hunk 3]
```diff
@@ -476,8 +522,15 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
           )}
         />
 
-        <Button type="submit" className="w-full">
-          {t("create_user")}
+        <Button
+          type="submit"
+          className="w-full"
+          disabled={
+            !form.formState.isDirty || !form.formState.isValid || isPending
```
Comment (AdityaJ2305) [inline src/components/Users/CreateUserForm.tsx]: @rajku-dev , `!form.formState.isValid` do we need it ? Zod don't let it submit until all validations met. Could you please confirm ?

Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: That is to avoid over populating the form with errors(frontend) when the user submits an invalid form
Changing the mode to `onChange` allows the users to see and address errors in real time.
Just a good UX that's all.
 did you try resolving the `.refine` part?

**Comment by nihal467** [comment]
@rajku-dev what is the status of this PR, we need to get it wrapped ASAP

**Comment by nihal467** [comment]
image

- [ ] show proper error message for unique email id error message from backend

**Comment by rajku-dev** [comment]
> image
> 
> * [ ]  show proper error message for unique email id error message from backend

It is already showing the backend error, Could you please confirm if there's any specific issue?
image

**Comment by rajku-dev** [comment]
> @rajku-dev what is the status of this PR, we need to get it wrapped ASAP

most of the things are already done.

File: src/components/Users/CreateUserForm.tsx [hunk 4]
```diff
@@ -66,23 +69,42 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
-      doctor_medical_council_registration: z.string().optional(),
-      geo_organization: z.string().min(1, t("field_required")),
+      date_of_birth: z
+        .date({
+          required_error: t("this_field_is_required"),
+        })
+        .refine((dob) => dob <= new Date(), {
+          message: t("date_of_birth_cannot_be_in_future"),
+        }),
+      gender: z.enum(["male", "female", "transgender", "non_binary"]),
```
[the remark this answers is not in the document]
Comment (bodhish) [reply src/components/Users/CreateUserForm.tsx]: This should be done right? @rajku-dev any reason for ignoring coderabbit. ?

File: src/components/Users/CreateUserForm.tsx [hunk 5]
```diff
@@ -66,23 +68,42 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
         .regex(/[A-Z]/, t("password_uppercase_validation"))
         .regex(/[0-9]/, t("password_number_validation")),
       c_password: z.string(),
-      first_name: z.string().min(1, t("field_required")),
-      last_name: z.string().min(1, t("field_required")),
-      email: z.string().email(t("invalid_email_address")),
+      first_name: z.string().min(1, t("this_field_is_required")),
+      last_name: z.string().min(1, t("this_field_is_required")),
```
Comment (bodhish) [inline src/components/Users/CreateUserForm.tsx]: Why make these changes? 🤔

File: src/components/Users/CreateUserForm.tsx [hunk 6]
```diff
@@ -66,23 +68,42 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
       phone_number_is_whatsapp: z.boolean().default(true),
-      date_of_birth: z.string().min(1, t("field_required")),
-      gender: z.enum(["male", "female", "other"]),
-      qualification: z.string().optional(),
-      doctor_experience_commenced_on: z.string().optional(),
-      doctor_medical_council_registration: z.string().optional(),
-      geo_organization: z.string().min(1, t("field_required")),
+      date_of_birth: z
+        .string({
+          required_error: t("this_field_is_required"),
```
Comment (bodhish) [inline src/components/Users/CreateUserForm.tsx]: What change does it make? `field_required` vs `this_field_is_required`

File: src/components/Users/CreateUserForm.tsx [hunk 7]
```diff
@@ -114,7 +138,14 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
       form.trigger("username");
     }
-  }, [phoneNumber, isWhatsApp, form, usernameInput]);
+    if (password?.length && c_password?.length && password !== c_password) {
+      form.setError("c_password", { message: t("password_mismatch") });
+      form.setError("password", { message: t("password_mismatch") });
+    } else {
+      form.clearErrors("c_password");
+      form.clearErrors("password");
+    }
```
Comment (bodhish) [inline src/components/Users/CreateUserForm.tsx]: Why are we doing this in useEffect, why not make it part of the onChange for the form fields🤔

Also since we are using `zod` why can't we use somehting like [superrefine](https://github.com/colinhacks/zod?tab=readme-ov-file#superrefine)

File: src/components/Users/CreateUserForm.tsx [hunk 8]
```diff
@@ -160,29 +191,39 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
+      date_of_birth: data.date_of_birth,
+      geo_organization: data.geo_organization,
+      user_type: data.user_type,
+      gender: data.gender,
+      password: data.password,
+      qualification: data.qualification,
+      doctor_experience_commenced_on: data.doctor_experience_commenced_on,
+      doctor_medical_council_registration:
+        data.doctor_medical_council_registration,
+    });
```
Comment (bodhish) [inline src/components/Users/CreateUserForm.tsx:227]: why can't we pass `createUser(data)` 🤔

**Review by bodhish — CHANGES_REQUESTED** [review]
What is happening here?
image

File: src/components/Users/CreateUserForm.tsx (hunk 8 above)
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx:227]: because data contains fields like `c_password` and `phone_number_is_whatsapp` that are only required for frontend and not required to send to backend
 `c_password` -> for frontend password match validation (not required in backend)
 `phone_number_is_whatsapp` -> to fill/unfill the `alt_phone_number` field

File: src/components/Users/CreateUserForm.tsx (hunk 5 above)
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: i first wrote translations for all error messages, then re-used the existing translations, probably missed that one, will make the changes

**Comment by rajku-dev** [comment]
> What is happening here? image

tbh, merge conflicts kind of scared me, especially when diffs were large, so i created new pr
would learn how to solve big merge conflicts.

File: src/components/Users/CreateUserForm.tsx (hunk 7 above)
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: although there was password mismatch error, the formState.error object is logging to be empty and the formState.isValid state to be false at the same time, which is causing the problem
there is also problem debugging the changes on local because access token lifetime is set to 5 mins i think.

File: src/components/Users/CreateUserForm.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: could not apply because @coderabbitai suggestion was wrong here.
image
but i would still consider coderabbit's suggestion in future.

[the remark this answers is not in the document]
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: applied ✅

File: src/components/Users/CreateUserForm.tsx [hunk 9]
```diff
@@ -68,20 +70,43 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
-      date_of_birth: z.string().min(1, t("field_required")),
-      gender: z.enum(["male", "female", "other"]),
-      qualification: z.string().optional(),
-      doctor_experience_commenced_on: z.string().optional(),
-      doctor_medical_council_registration: z.string().optional(),
+      date_of_birth: z
+        .string({
+          required_error: t("field_required"),
+        })
+        .refine((dob) => dob <= new Date().toISOString(), {
```
Comment (rithviknishad) [inline src/components/Users/CreateUserForm.tsx]: didn't know strings could compared

File: src/components/Users/CreateUserForm.tsx (hunk 7 above)
Comment (rithviknishad) [reply src/components/Users/CreateUserForm.tsx]: access token should ideally get refreshed automatically behind the scenes. was it not happening?

Comment (rithviknishad) [reply src/components/Users/CreateUserForm.tsx]: > although there was password mismatch error, the formState.error object is logging to be empty and the formState.isValid state to be false at the same time, which is causing the problem

☝🏼 i don't think this explains anything why refine doesn't work

File: src/components/Users/CreateUserForm.tsx [hunk 10]
```diff
@@ -68,30 +70,58 @@
...
+          message: t("date_of_birth_cannot_be_in_future"),
+        }),
```
[the remark this answers is not in the document]
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: @coderabbitai make sure you also suggest new translation key-value pair in `en.json` for new translations used in suggestion, if it does not exist in `en.json`

File: src/components/Users/CreateUserForm.tsx (hunk 9 above)
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: applied coderabbit suggestion.

File: src/components/Users/CreateUserForm.tsx [hunk 11]
```diff
@@ -114,7 +146,14 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
       form.trigger("username");
     }
-  }, [phoneNumber, isWhatsApp, form, usernameInput]);
+    if (password?.length && c_password?.length && password !== c_password) {
+      form.setError("c_password", { message: t("password_mismatch") });
+      form.setError("password", { message: t("password_mismatch") });
+    } else {
+      form.clearErrors("c_password");
+      form.clearErrors("password");
+    }
```
[the remark this answers is not in the document]
Comment (rajku-dev) [reply src/components/Users/CreateUserForm.tsx]: why form.formState.isValid = false and form.formState.errors object empty at the same time.

File: src/components/Users/CreateUserForm.tsx [hunk 12]
```diff
@@ -68,20 +70,51 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
...
+      date_of_birth: z
+        .string({
+          required_error: t("field_required"),
+        })
+        .refine(
+          (dob) => {
+            const date = new Date(dob);
+            const minDate = new Date();
+            minDate.setFullYear(minDate.getFullYear() - 120);
+            return date >= minDate && date <= new Date();
```
Comment (rithviknishad) [inline src/components/Users/CreateUserForm.tsx]: nitpick, just easier to read/comprehend comparisons from Math perspective: `a < x < b` kinda style.

```suggestion
            return minDate <= date && date <= new Date();
```

**Comment by rithviknishad** [comment]
@rajku-dev request for review only once previous reviews has been resolved

**Comment by rajku-dev** [comment]
> @rajku-dev request for review only once previous reviews has been resolved

would keep this in mind, apologies

**Comment by rajku-dev** [comment]
removed the `isValid` check on Submit button

**Comment by nihal467** [comment]
@rajku-dev use proper PR heading

**Comment by nihal467** [comment]
LGTM @rithviknishad can you review it

**Review by rithviknishad — APPROVED** [review]
LGTM; types changes requires review from @bodhish

File: src/types/user/user.ts [hunk 13]
```diff
@@ -14,3 +14,20 @@ export type UserBase = {
...
+  username: string;
+  first_name: string;
+  last_name: string;
+  email: string;
+  phone_number: string;
+  alt_phone_number?: string;
+  date_of_birth: string;
+  qualification?: string;
+  doctor_experience_commenced_on?: string;
+  doctor_medical_council_registration?: string;
```
Comment (Jacobjeevan) [inline src/types/user/user.ts:32]: Hmm 🤔  I'm also adding something similar here: https://github.com/ohcnetwork/care_fe/pull/10027

cc: @rithviknishad 

DoB is going away btw.

**Comment by rajku-dev** [comment]
Hi @Jacobjeevan , are these changes good to go with #10027 
It's been quite some time since this is open.

**Comment by Jacobjeevan** [comment]
> Hi @Jacobjeevan , are these changes good to go with #10027 It's been quite some time since this is open.

Hmm, many of the changes are not needed since the fields are not relevant 🤔  I'm going to close this PR and include any relevant changes in the other PR (marking the required fields); I'll add you as a co-author for the commit.

**Comment by rajku-dev** [comment]
> > Hi @Jacobjeevan , are these changes good to go with #10027 It's been quite some time since this is open.
> 
> Hmm, many of the changes are not needed since the fields are not relevant 🤔 I'm going to close this PR and include any relevant changes in the other PR (marking the required fields); I'll add you as a co-author for the commit.

That's ok 🙂
